yunwoo/programmers-2/[3차]-파일명-정렬.js

```diff
@@ -0,0 +1,38 @@
+function solution(files) {
+  const inputObjArr = [];
+  const reg = /[0-9]+/;
+
+  files.forEach((file) => {
+    const number = file.match(reg)[0];
+    const numberIdx = file.indexOf(number);
+    const head = file.substring(0, numberIdx);
+    const tail = file.substring(numberIdx + number.length, file.length);
+
+    inputObjArr.push({ head, number, tail });
+  });
+
+  inputObjArr.sort((a, b) => {
+    const aHead = a.head.toLowerCase();
+    const bHead = b.head.toLowerCase();
+
+    if (bHead > aHead) return -1;
+    if (bHead < aHead) return 1;
+
+    return Number(a.number) - Number(b.number);
+  });
+
+  return inputObjArr.reduce((acc, cur) => {
+    const joinInputObj = `${cur.head}${cur.number}${cur.tail}`;
+
+    acc.push(joinInputObj);
+
+    return acc;
+  }, []);
+}
```

Comment:
```suggestion
function solution(files) {
  const reg = /[0-9]+/;

  const mappingFiles = files.map((file) => {
    const number = file.match(reg)[0];
    const numberIdx = file.indexOf(number);
    const head = file.substring(0, numberIdx).toLowerCase();

    return { origin: file, head, number };
  });

  return mappingFiles
    .sort((a, b) => {
      const aNum = +a.number;
      const bNum = +b.number;

      return a.head.localeCompare(b.head) || aNum - bNum;
    })
    .map((file) => file.origin);
}
```
- 풀이잘봤습니다!
- forEach를 이용하기보다 map을 이용해서 객체요소를 가진 배열로 핸들링 후에 이를 활용하면 좀 더 깔끔한 풀이가 가능합니다! 이때 원본 파일 이름을 같이 넘겨주면 더욱 좋습니다!
- +사실 해당 문제는 tail이 필요 없습니다!
- sort 시에 OR연산자를 활용하면 더욱 깔끔한 코드가 가능합니다.

Reply:
- 아!! 원본을 어떻게 유지하지 고민했었는데 그냥 map을 돌리고 원본을 넘겨주면 되는군요ㅎㅎ..😅
- localeCompare 메서드를 안쓴지 오래되서 잊고 있었네요 👍
- OR연산자로 0일 경우에는 동일하니까 넘버로 넘어가는군요!! 깔끔한 풀이 감사합니다 🙇‍♂️